google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -3343,6 +3441,43 @@ private void createSessions(final int sessionCount, boolean distributeOverChanne
     }
   }
 
+  class MultiplexedSessionConsumer implements SessionConsumer {
+
+    @Override
+    public void onSessionReady(SessionImpl sessionImpl) {
+      final SettableFuture settableFuture = SettableFuture.create();
+      final MultiplexedSession newSession = new MultiplexedSession(sessionImpl);
+      settableFuture.set(newSession);
+
+      synchronized (lock) {
+        if (multiplexedSessionFuture.get() != null && multiplexedSessionFuture.get().isDone()) {
+          final MultiplexedSession oldSession = getMultiplexedSession().get();
+          logger.log(
+              Level.INFO,
+              String.format(
+                  "Removed Multiplexed Session => %s created at => %s and",
+                  oldSession.getName(), oldSession.getDelegate().getCreateTime()));
+          if (multiplexedSessionRemovedListener != null) {
+            multiplexedSessionRemovedListener.apply(oldSession);
+          }
```

Comment:
nit: could the code block for logging and notifying about the removal of the old session to after it has actually been removed? The reason that I think that would be better, is that I think that the code block that hands out the multiplexed session should not take the `lock` (see comment further above). That means that getting a session could overlap with this code block that has taken the lock, and that you could (very theoretically) get a situation where the `removedListener` is called for a session first, and then that the session that was notified as 'removed' was still handed out by the pool.

Reply:
Yes, this was actually causing a flaky behavior in unit test `MultiplexedSessionMaintainerTest`. I have fixed it.